crates/findex/src/implementations/redis/findex.rs

```diff
@@ -0,0 +1,434 @@
+        FindexTable::Entry => b"fe::".to_vec(),
+        FindexTable::Chain => b"fc::".to_vec(),
+    };
+    key.extend_from_slice(uid);
+    key
+}
+
+pub struct FindexRedis {
+    // we keep redis_url for the updateLines method
+    mgr: ConnectionManager,
+    upsert_script: Script,
+    removed_locations_finder: Arc<dyn RemovedLocationsFinder + Sync + Send>,
+    compact_lock: RwLock<()>,
+}
+
+impl FindexRedis {
+    /// The conditional upsert script used to
+    /// only update a table if the previous value matches ARGV[2].
+    /// When the value does not match, the previous value is returned
+    const CONDITIONAL_UPSERT_SCRIPT: &str = r#"
+        local value=redis.call('GET',ARGV[1])
+        if((value==false) or (not(value == false) and (ARGV[2] == value))) then 
+            redis.call('SET', ARGV[1], ARGV[3])
+            return  
+        else 
+            return value 
+        end;
+    "#;
+
+    /// Connect to a Redis server
+    ///
+    /// # Arguments
+    ///  * `redis_url` - The Redis URL e.g.
+    ///    "redis://user:password@localhost:6379"
+    pub async fn connect(
+        redis_url: &str,
+        removed_locations_finder: Arc<dyn RemovedLocationsFinder + Sync + Send>,
+    ) -> Result<Self, FindexRedisError> {
+        let client = redis::Client::open(redis_url)?;
+        let mgr = ConnectionManager::new(client).await?;
+
+        Ok(FindexRedis {
+            mgr,
+            upsert_script: Script::new(Self::CONDITIONAL_UPSERT_SCRIPT),
+            removed_locations_finder,
+            compact_lock: RwLock::new(()),
+        })
+    }
+
+    /// Connect to a Redis server with a ConnectionManager
+    ///
+    /// # Arguments
+    ///  * `redis_url` - The Redis URL e.g.
+    ///    "redis://user:password@localhost:6379"
+    pub async fn connect_with_manager(
+        mgr: ConnectionManager,
+        removed_locations_finder: Arc<dyn RemovedLocationsFinder + Sync + Send>,
+    ) -> Result<Self, FindexRedisError> {
+        Ok(FindexRedis {
+            mgr,
+            upsert_script: Script::new(Self::CONDITIONAL_UPSERT_SCRIPT),
+            removed_locations_finder,
+            compact_lock: RwLock::new(()),
+        })
+    }
+
+    /// Clear all indexes
+    ///
+    /// # Warning
+    /// This is definitive
+    pub async fn clear_indexes(&self) -> Result<(), FindexRedisError> {
+        redis::cmd("FLUSHDB")
+            .query_async(&mut self.mgr.clone())
+            .await?;
+        Ok(())
+    }
+
+    /// Upsert the given chain elements in Findex tables.
+    ///
+    /// # Parameters
+    ///
+    /// - `master_key`  : Findex master key
+    /// - `label`       : additional public information used in key hashing
+    /// - `additions`   : values to indexed for a set of keywords
+    /// - `deletions`   : values to remove from the indexes for a set of
+    ///   keywords
+    ///
+    /// # Returns
+    /// The set of new keywords that were added
+    pub async fn upsert(
```

Comment:
Can you explain the rational behind these Findex API wrappers?

Reply:
Yes, it is a common pattern called a "facade". It does two things:
 - avoid the user importing traits e.g., `FindexSearch`, `FindexUpdate`, etc...
 - avoid exposing the user to public methods in those traits that it should not be using
 
 As a general rule, implementations should always provide facades to complex APIs, to simplify the user API